File: UI/src/services/GpaoService.ts
import env from '../config/env';
import authService from './AuthService';

const API_BASE_URL = env.BASE_URL + '/api/gpao';

class GpaoService {
    private async get(endpoint: string, requireAuth: boolean = true) {
        const headers: Record<string, string> = {
            'Content-Type': 'application/json',
        };

        // Ajout du token uniquement si requis
        if (requireAuth) {
            const token = authService.getToken();
            if (!token) throw new Error("Utilisateur non authentifié");
            headers['Authorization'] = `Bearer ${token}`;
        }

        const response = await fetch(`${API_BASE_URL}${endpoint}`, {
            method: 'GET',
            headers,
        });

        if (!response.ok) {
            const errorData = await response.json().catch(() => ({ message: `Request failed with status ${response.status}` }));
            throw new Error(errorData.message || 'API request failed');
        }

        return response.json();
    }

    private async post(endpoint: string, body: object, requireAuth: boolean = true) {
        const headers: Record<string, string> = {
            'Content-Type': 'application/json',
        };

        // Ajout du token uniquement si requis
        if (requireAuth) {
            const token = authService.getToken();
            if (!token) throw new Error("Utilisateur non authentifié");
            headers['Authorization'] = `Bearer ${token}`;
        }

        const response = await fetch(`${API_BASE_URL}${endpoint}`, {
            method: 'POST',
            headers,
            body: JSON.stringify(body),
        });

        if (!response.ok) {
            const errorData = await response.json().catch(() => ({ message: `Request failed with status ${response.status}` }));
            throw new Error(errorData.message || 'API request failed');
        }

        const contentType = response.headers.get("content-type");
        if (contentType && contentType.indexOf("application/json") !== -1) {
            return response.json();
        } else {
            return {};
        }
    }

    // === 🟢 METHODES GPAO ===

    /**
     * Récupère les lots client d’un dossier (aucune authentification requise)
     */
    getClientLots(idDossier: number) {
        return this.post('/lot/get-client-lots', { idDossier }, false);
    }

    getCurrentLotForUser(idDossier: number, idEtape: number, idPers: number, idLotClient: number) {
        return this.post('/lot/get-lot', {
            _idDossier: idDossier,
            _idEtape: idEtape,
            _idPers: idPers,
            _idLotClient: idLotClient,
        });
    }

    getLot(idDossier: number, idEtape: number, idPers: number, idLotClient: number) {
        return this.post('/lot/get-lot', {
            _idDossier: idDossier,
            _idEtape: idEtape,
            _idPers: idPers,
            _idLotClient: idLotClient,
        });
    }

    startNewLdt(params: { _idDossier: number; _idEtape: number; _idPers: number; _idLotClient: number; _idLot: number; _idTypeLdt: number; }) {
        return this.post('/ldt/start', params);
    }
    
    endLdt(params: { _idDossier: number; _idEtape: number; _idPers: number; _idLotClient: number; _idLot: number; _idTypeLdt: number; _qte: number; }) {
        return this.post('/ldt/end', params);
    }

    injectNextEtape(params: { _idDossier: number; _idNextEtape: number; _idLotClient: number; _libelle: string; _qte: number; }) {
        return this.post('/lot/inject-next-etape', params);
    }

    updateLot(params: { _idLot: number; _idEtat: number; _qte: number; }) {
        return this.post('/lot/update-lot', params);
    }
}

export default new GpaoService();

File: WEB_SERVICE/services/gpao_service.py
import requests
import os
from WEB_SERVICE.models.lot_model import Lot
from WEB_SERVICE.utils.constant import BASE_PATH, BASE_URL_API_GPAO, ID_ETAPE_CQ_ISO

class GpaoService:
    
    @staticmethod
    def post(endpoint: str, payload: dict, token: str):
        """
        Méthode générique pour poster vers l'API GPAO externe.
        Utilise les paramètres en query string au lieu du body JSON.
        """
        url = f"{BASE_URL_API_GPAO}{endpoint}"
        headers = {
            "Authorization": f"Bearer {token}"
        }
        
        try:
            # Envoi en query params
            response = requests.post(url, params=payload, headers=headers, timeout=10)

            if response.status_code != 200:
                print(f"[GpaoService] Erreur {response.status_code} sur {endpoint} : {response.text}")
                return None

            # Retourne le JSON si possible
            try:
                json_data = response.json()
                print(f"[GpaoService] Réponse reçue de {endpoint} : {json_data}")
                return json_data
            except ValueError:
                return {"status": "ok"}  # Réponse vide mais succès

        except requests.exceptions.RequestException as e:
            print(f"[GpaoService] Exception lors de l'appel {endpoint} : {e}")
            return None

    # === Fonctions métiers GPAO ===
    @staticmethod
    def get_lot(data: dict, token: str):
        result = GpaoService.post("/Lot/get-lot", data, token)

        if not result or "lot" not in result:
            return result

        lot = result["lot"]
        client = lot.get("libelleLotClient")
        libelle = lot.get("libelle")
        libelle_without_extension = ""
        print(f"[GpaoService] Traitement du lot : {lot}")
        
        # Vérifier et enlever l'extension .tif si elle existe
        if libelle and libelle.lower().endswith(".tif"):
            libelle_without_extension = os.path.splitext(libelle)[0]

        # Construction des chemins
        lot["paths"] = {
            "basePath": BASE_PATH,
            "IN_CQ": f"{BASE_PATH}/SAISIE/{client}/{libelle_without_extension}/{libelle_without_extension}.csv",
            "OUT_CQ": f"{BASE_PATH}/CQ_CIBLE/{client}/{libelle_without_extension}",   # dossier uniquement
            "IMAGE_PATH": f"{BASE_PATH}/IMAGE/{client}/{libelle}",
            "IN_CQ_ISO": f"{BASE_PATH}/CQ_CIBLE/{client}/{libelle_without_extension}/{libelle_without_extension}.csv",
            "OUT_CQ_ISO": f"{BASE_PATH}/CQ_ISO/{client}/{libelle_without_extension}",
        }

        
        # Recherche de l'utilisateur en base
        lotGpao = Lot.query.filter_by(id_lot=lot.get("idLot")).first()
        if not lotGpao:
            print(f"[GpaoService] Lot {lot.get('idLot')} introuvable dans p_lot")
            return None
        
        # ajouter idEtape depuis p_lot
        lot["idEtape"] = lotGpao.id_etape
        lot["idNextEtape"] = ID_ETAPE_CQ_ISO  # Préparer pour l'injection dans l'étape suivante (idEtape CQ_ISO)

        return result

    @staticmethod
    def start_ldt(data: dict, token: str):
        return GpaoService.post("/Ldt/start", data, token)

    @staticmethod
    def end_ldt(data: dict, token: str):
        return GpaoService.post("/Ldt/end", data, token)

    @staticmethod
    def inject_next_etape(data: dict, token: str):
        return GpaoService.post("/Lot/inject-next-etape", data, token)

    @staticmethod
    def update_lot(data: dict, token: str):
        return GpaoService.post("/Lot/update-lot", data, token)
